htlcswitch/quiescer.go

```diff
@@ -0,0 +1,83 @@
+package htlcswitch
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// quiescer is a state machine that tracks progression through the quiescence
+// protocol.
+type quiescer struct {
+	chanID   lnwire.ChannelID
+	sent     bool
+	received bool
+}
+
+// recvStfu is called when we receive an Stfu message from the remote. It
+// returns a bool indicating whether or not the channel is now quiescent as a
+// result.
+func (q *quiescer) recvStfu() (bool, error) {
+	if q.received {
+		return false,
+			fmt.Errorf(
+				"stfu already received for channel %s",
```

Comment:
If we promote these to either an error variable or a simple error struct, then unit tests can use `errors.As` and `errors.Is`, etc. 

The caller/driver of the state machine would then also be able to use a switch to handle the error case.